airflow/jobs.py

```diff
@@ -1891,112 +1940,126 @@ def _manage_executor_state(self, started):
                     self.logger.error(msg)
                     ti.handle_failure(msg)
 
-    def _execute(self):
+    @provide_session
+    def _get_dag_run(self, run_date, session=None):
         """
-        Runs a dag for a specified date range.
+        Returns a dag run for the given run date, which will be matched to an existing
+        dag run if available or create a new dag run otherwise.
+        :param run_date: the execution date for the dag run
+        :param run_date: datetime
+        :param session:
+        :return:
         """
-        session = settings.Session()
-        DagRun = models.DagRun
-
-        # consider max_active_runs but ignore when running subdags
-        # "parent.child" as a dag_id is by convention a subdag
-        if self.dag.schedule_interval and not self.dag.is_subdag:
-            active_runs = DagRun.find(
-                dag_id=self.dag.dag_id,
+        run_id = BackfillJob.ID_FORMAT_PREFIX.format(run_date.isoformat())
+
+        # check if we are scheduling on top of a already existing dag_run
+        # we could find a "scheduled" run instead of a "backfill"
+        run = DagRun.find(dag_id=self.dag.dag_id,
+                          execution_date=run_date,
+                          session=session)
+
+        if not run:
+            run = self.dag.create_dagrun(
+                run_id=run_id,
+                execution_date=run_date,
+                start_date=datetime.now(),
                 state=State.RUNNING,
                 external_trigger=False,
                 session=session
             )
+        else:
+            run = run[0]
 
-            # return if already reached maximum active runs
-            if len(active_runs) >= self.dag.max_active_runs:
-                self.logger.info("Dag {} has reached maximum amount of {} dag runs"
-                                 .format(self.dag.dag_id, self.dag.max_active_runs))
-                return
-
-        start_date = self.bf_start_date
-        end_date = self.bf_end_date
+        # set required transient field
+        run.dag = self.dag
 
-        # picklin'
-        pickle_id = None
-        if not self.donot_pickle and self.executor.__class__ not in (
-                executors.LocalExecutor, executors.SequentialExecutor):
-            pickle = models.DagPickle(self.dag)
-            session.add(pickle)
-            session.commit()
-            pickle_id = pickle.id
-
-        executor = self.executor
-        executor.start()
+        # explicitly mark as backfill and running as we can fill gaps
```

Comment:
What does "as we can fill gaps" mean? I see it's copied from the old logic, let's kill the second part of this sentence if you don't know what it means either.